lib/BuildSystem/BuildSystem.cpp

```diff
@@ -885,6 +887,123 @@ class DirectoryContentsTask : public Task {
 };
 
 
+/// This task is responsible for computing the filtered lists of files in
+/// directories.
+class FilteredDirectoryContentsTask : public Task {
+  std::string path;
+
+  /// The exclusion filters used while computing the signature
+  StringList filters;
+
+  /// The value for the input directory.
+  BuildValue directoryValue;
+
+  virtual void start(BuildEngine& engine) override {
+    // Depend on the node so that it can be connected up to potential producers.
+    engine.taskNeedsInput(
+        this, BuildKey::makeNode(path).toData(), /*inputID=*/0);
+    // Depend on the raw stat information, because we want to know if we need
+    // to rescan items under this if anything has changed.
+    engine.taskNeedsInput(
+        this, BuildKey::makeStat(path).toData(), /*inputID=*/1);
+  }
+
+  virtual void providePriorValue(BuildEngine&,
+                                 const ValueType& value) override {
+  }
+
+  virtual void provideValue(BuildEngine&, uintptr_t inputID,
+                            const ValueType& value) override {
+    // If our value is uninitialized, take whatever build value came in
+    if (directoryValue.isInvalid()) {
+      directoryValue = BuildValue::fromData(value);
+      return;
+    }
+
+    // If our value doesn't exist yet, take subsequent input
+    if (!directoryValue.isExistingInput()) {
+      directoryValue = BuildValue::fromData(value);
+      return;
+    }
+  }
+
+  virtual void inputsAvailable(BuildEngine& engine) override {
+    if (directoryValue.isMissingInput()) {
+      engine.taskIsComplete(this, BuildValue::makeMissingInput().toData());
+      return;
+    }
+
+    if (!directoryValue.isExistingInput()) {
+      engine.taskIsComplete(this, BuildValue::makeFailedInput().toData());
+      return;
+    }
+
+    auto& info = directoryValue.getOutputInfo();
+
+    // Non-directory things are just plain-ol' inputs
+    if (!info.isDirectory()) {
+      engine.taskIsComplete(this, BuildValue::makeExistingInput(info).toData());
+      return;
+    }
+
+    // Collec the filtered contents
+    std::vector<std::string> filenames;
+    getFilteredContents(path, filters, filenames);
+
+    // Compute the filters signature
+    using llvm::hash_combine;
+    llvm::hash_code code = hash_value(path);
+    for (const auto& filter: filters.getValues()) {
+      code = hash_combine(code,
+                          hash_combine_range(filter.begin(), filter.end()));
+    }
```

Comment:
I don't think we actually need a signature for the filters. Why would we? The filters are a part of the build key?

Reply:
Ah, fair point.